test/extended/single_node/topology.go

```diff
@@ -0,0 +1,149 @@
+	"context"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	v1 "github.com/openshift/api/config/v1"
+	exutil "github.com/openshift/origin/test/extended/util"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	e2e "k8s.io/kubernetes/test/e2e/framework"
+	e2eskipper "k8s.io/kubernetes/test/e2e/framework/skipper"
+	"strings"
+)
+
+func getOpenshiftNamespaces(f *e2e.Framework) []corev1.Namespace {
+	list, err := f.ClientSet.CoreV1().Namespaces().List(context.Background(), metav1.ListOptions{})
+	Expect(err).NotTo(HaveOccurred())
+
+	var openshiftNamespaces []corev1.Namespace
+	for _, namespace := range list.Items {
+		if strings.HasPrefix(namespace.Name, "openshift-") {
+			openshiftNamespaces = append(openshiftNamespaces, namespace)
+		}
+	}
+
+	return openshiftNamespaces
+}
+
+func getNamespaceDeployments(f *e2e.Framework, namespace corev1.Namespace) []appsv1.Deployment {
+	list, err := f.ClientSet.AppsV1().Deployments(namespace.Name).List(context.Background(), metav1.ListOptions{})
+	Expect(err).NotTo(HaveOccurred())
+
+	var namespaceDeployments []appsv1.Deployment
+	for _, namespace := range list.Items {
+		namespaceDeployments = append(namespaceDeployments, namespace)
+	}
+
+	return namespaceDeployments
+}
+
+func getTopologies(f *e2e.Framework) (controlPlaneTopology, infraTopology v1.TopologyMode) {
+	oc := exutil.NewCLIWithFramework(f)
+	infra, err := oc.AdminConfigClient().ConfigV1().Infrastructures().Get(context.Background(),
+		"cluster", metav1.GetOptions{})
+	Expect(err).NotTo(HaveOccurred())
+
+	return infra.Status.ControlPlaneTopology, infra.Status.ControlPlaneTopology
+}
+
+func isInfrastructureDeployment(deployment appsv1.Deployment) bool {
+	infrastructureNamespaces := map[string][]string{
+		"openshift-ingress": {
+			"router-default",
+		},
+	}
+
+	namespaceInfraDeployments, ok := infrastructureNamespaces[deployment.Namespace]
+
+	if !ok {
+		return false
+	}
+
+	for _, infraDeploymentName := range namespaceInfraDeployments {
+		if deployment.Name == infraDeploymentName {
+			return true
+		}
+	}
+
+	return false
+}
+
+func validateReplicas(deployment appsv1.Deployment,
+	controlPlaneTopology, infraTopology v1.TopologyMode, failureAllowed bool) {
+	if isInfrastructureDeployment(deployment) {
+		if infraTopology != v1.SingleReplicaTopologyMode {
+			return
+		}
+	} else if controlPlaneTopology != v1.SingleReplicaTopologyMode {
+		return
+	}
+
+	Expect(deployment.Spec.Replicas).ToNot(BeNil())
+
+	replicas := int(*deployment.Spec.Replicas)
+
+	if !failureAllowed {
+		Expect(replicas).To(Equal(1),
+			"%s in %s namespace has wrong number of replicas", deployment.Name, deployment.Namespace)
+	} else {
+		if replicas == 1 {
+			t := GinkgoT()
+			t.Logf("Deployment %s in namespace %s has one replica, consider taking it off the topology allow-list",
+				deployment.Name, deployment.Namespace)
+		}
+	}
+}
+
+func isAllowedToFail(deployment appsv1.Deployment) bool {
```

Comment:
A comment explaining these special cases would be useful for future reference

Reply:
Sure. Will add one

Reply:
Added in 395d46965a